File: WL/HLControl/HLRadioButton.vb
﻿Namespace HLControl

    <DefaultEvent("CheckedChanged")>
    Public Class HLRadioButton
        Inherits Control

        Private 值 As Boolean, 行高 As Integer

        Public Sub New()
            DoubleBuffered = True
            值 = False
        End Sub

        Private Sub _NeedRePaint() Handles Me.SizeChanged, Me.Resize, Me.AutoSizeChanged, Me.FontChanged, Me.EnabledChanged
            Invalidate()
        End Sub

        Private Sub _MouseDown(sender As Object, e As MouseEventArgs) Handles Me.MouseDown
            Value = True
        End Sub

        <DefaultValue(False)>
        Public Property Value As Boolean
            Get
                Return 值
            End Get
            Set(v As Boolean)
                If v <> 值 Then
                    If 非空(Parent) Then
                        For Each i As Control In Parent.Controls
                            If i.GetType = [GetType]() AndAlso i.GetHashCode <> GetHashCode() Then
                                Dim r As HLRadioButton = i
                                r.Value = False
                            End If
                        Next
                    End If
                    值 = v
                    RaiseEvent CheckedChanged()
                    Invalidate()
                End If
            End Set
        End Property

        Public Event CheckedChanged()

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            修正Dock(Me, True, False)
            行高 = Font.GetHeight
            Height = 行高
            MyBase.OnPaint(e)
            Dim g As Graphics = e.Graphics, c As Rectangle = ClientRectangle
            With g
                Width = .MeasureString(Text, Font).Width + Height
                Dim h As Single = 行高 * 0.6
                Dim p As Single = (行高 - h) / 2
                .DrawEllipse(边缘灰笔, New RectangleF(p, p, h, h))
                h *= 0.6
                p = (行高 - h) / 2
                If Value AndAlso Enabled Then .FillEllipse(白色笔刷, New RectangleF(p, p, h, h))
                绘制文本(g, Text, Font, 行高, 0, 获取文本状态(Enabled, Value))
            End With
        End Sub

    End Class

End Namespace

File: decompilled2csharp/WL/WL.HLControl/HLRadioButton.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Windows.Forms;
using WL.基础;

namespace WL.HLControl
{
	[DefaultEvent("CheckedChanged")]
	public class HLRadioButton : HLControlBase
	{
		public delegate void CheckedChangedEventHandler(HLRadioButton sender, HLValueEventArgs e);

		private bool 值;

		private int 行高;

		[CompilerGenerated]
		[DebuggerBrowsable(DebuggerBrowsableState.Never)]
		private CheckedChangedEventHandler CheckedChangedEvent;

		[DefaultValue(false)]
		public bool Checked
		{
			get
			{
				return 值;
			}
			set
			{
				if (value != 值)
				{
					if (类型.非空(base.Parent))
					{
						IEnumerator enumerator = default(IEnumerator);
						try
						{
							enumerator = base.Parent.Controls.GetEnumerator();
							while (enumerator.MoveNext())
							{
								Control control = (Control)enumerator.Current;
								if (control.GetType() == GetType() && control.GetHashCode() != GetHashCode())
								{
									HLRadioButton hLRadioButton = (HLRadioButton)control;
									hLRadioButton.Checked = false;
								}
							}
						}
						finally
						{
							if (enumerator is IDisposable)
							{
								(enumerator as IDisposable).Dispose();
							}
						}
					}
					int num = 0 - (值 ? 1 : 0);
					值 = value;
					CheckedChangedEvent?.Invoke(this, new HLValueEventArgs(num, 值));
					Invalidate();
				}
			}
		}

		public event CheckedChangedEventHandler CheckedChanged
		{
			[CompilerGenerated]
			add
			{
				CheckedChangedEventHandler checkedChangedEventHandler = CheckedChangedEvent;
				CheckedChangedEventHandler checkedChangedEventHandler2;
				do
				{
					checkedChangedEventHandler2 = checkedChangedEventHandler;
					CheckedChangedEventHandler value2 = (CheckedChangedEventHandler)Delegate.Combine(checkedChangedEventHandler2, value);
					checkedChangedEventHandler = Interlocked.CompareExchange(ref CheckedChangedEvent, value2, checkedChangedEventHandler2);
				}
				while ((object)checkedChangedEventHandler != checkedChangedEventHandler2);
			}
			[CompilerGenerated]
			remove
			{
				CheckedChangedEventHandler checkedChangedEventHandler = CheckedChangedEvent;
				CheckedChangedEventHandler checkedChangedEventHandler2;
				do
				{
					checkedChangedEventHandler2 = checkedChangedEventHandler;
					CheckedChangedEventHandler value2 = (CheckedChangedEventHandler)Delegate.Remove(checkedChangedEventHandler2, value);
					checkedChangedEventHandler = Interlocked.CompareExchange(ref CheckedChangedEvent, value2, checkedChangedEventHandler2);
				}
				while ((object)checkedChangedEventHandler != checkedChangedEventHandler2);
			}
		}

		public HLRadioButton()
		{
			base.MouseDown += _MouseDown;
			DoubleBuffered = true;
			值 = false;
		}

		private void _MouseDown(object sender, MouseEventArgs e)
		{
			Checked = true;
		}

		protected override void OnPaint(PaintEventArgs e)
		{
			HL辅助信息.修正Dock(this);
			checked
			{
				行高 = (int)Math.Round(Font.GetHeight());
				base.Height = 行高;
				base.OnPaint(e);
				Graphics graphics = e.Graphics;
				Rectangle clientRectangle = base.ClientRectangle;
				Graphics graphics2 = graphics;
				base.Width = (int)Math.Round(graphics2.MeasureString(Text, Font).Width + (float)base.Height);
				float num = (float)((double)行高 * 0.6);
				float num2 = ((float)行高 - num) / 2f;
				graphics2.DrawEllipse(HL辅助信息.边缘灰笔, new RectangleF(num2, num2, num, num));
				num = (float)((double)num * 0.6);
				num2 = ((float)行高 - num) / 2f;
				if (Checked && base.Enabled)
				{
					graphics2.FillEllipse(HL辅助信息.白色笔刷, new RectangleF(num2, num2, num, num));
				}
				HL辅助信息.绘制文本(graphics, Text, Font, 行高, 0f, HL辅助信息.获取文本状态(base.Enabled, Checked));
				graphics2 = null;
			}
		}
	}
}
